Voicechannels

jespino opened the pull request:
#### Summary
Adding support for calls plugin inside the mobile application. Everything is behind a feature flag (that is in a PR pending to be merge in mattermost-server repo).

The rest of the application is in the mattermost-plugin-calls in a special branch that is under development now.

**WARNING**: This PR still needs some extra work on adding unit tests.

#### TODO
- [x] Add tests
- [x] Use the last version of compass-icons

#### Outstanding pending bugs

- [ ]    If connection is lost it does not attempt to reconnect to the current call (should check if the current call is still available)
- [x]    The Floating banners are hiding the No internet connection bar and the more messages above bar
- [x]    When I press join a call without internet connection the UI seems to connect it, the mic does not do anything, can actually join the call, then it goes into this very very odd behavior.
- [x]    ~~If the call ends post does not seem to update after losing connection~~ This shouldn't happen as part of the application flow, it is related to the post props from the `custom_calls` post type. So... the server should update the post, and the mobile app should get that information available in the redux store.
- [x]    Persisting data can make it so a Join call banner is displayed when there is no ongoing call and the request to fetch the call data fails
- [x] I can go into the channel info screen and start as many calls as I want regardless if there is one in progress or not

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Has UI changes
- [x] Includes text changes and localization file updates

#### Device Information
This PR was tested on: Android

#### Screenshots
I'll add them later

#### Release Note
```release-note
NONE
```

mattermod commented:
Building app in separate branch.

mattermod commented:
Artifact links:  
https://18550-70265724-gh.circle-artifacts.com/0/tmp/artifacts/Mattermost_Beta.ipa  
https://18549-70265724-gh.circle-artifacts.com/0/tmp/artifacts/Mattermost_Beta.apk

jespino commented:
@enahum I addressed almost everything, I think we can go ahead if you find it right, I still have to do some manual testing, and I will iterate later on the re-connect problem, but other than that, I think I addressed almost everything.

enahum commented on app/components/post_draft/quick_actions/file_quick_action/index.tsx:
```diff
@@ -83,7 +83,7 @@ const FileQuickAction = ({disabled, fileCount = 0, intl, maxFileCount, onUploadF
             <CompassIcon
                 color={color}
-                name='file-document-outline'
+                name='file-text-outline'
```
@matthewbirtch it seems the `file-document-outline` is no longer included in compassIcon, @jespino made this change and for this PR is fine.. but I want to make sure this:
1. works fine for now
2. We submit a PR to use the right icon for this

matthewbirtch replied:
yes, sorry, the name changed to `file-text-outline`

jespino replied:
Great, looks like the change is right :)

enahum commented on app/components/post_list/more_messages_button/more_messages_button.tsx:
```diff
@@ -169,18 +189,22 @@ export default class MoreMessageButton extends React.PureComponent<MoreMessagesB
         }
     }
 
+    getBarsFactor = () => {
```
@jespino if I understand this right, the blue bar that allows you to scroll to the new message line indicator will display below the call indicator. yes ?

jespino replied:
Yes, bellow the indicator, or bellow the indicator plus the current call floating widget.

enahum reviewed (approved):
For the most part it looks good.. there are some comments that I included to make it a bit better but 0/5 on those.

and there are a couple of UX/UI questions

jespino commented on app/products/calls/components/join_call/join_call.tsx:
```diff
@@ -0,0 +1,129 @@
+            </Text>
+            <View style={style.avatars}>
+                <Avatars
+                    userIds={Object.values(props.call.participants || {}).map((x) => x.id)}
```
good call

jespino replied:
Done

jespino replied:
done

jespino commented on app/store/index.ts:
```diff
@@ -79,6 +79,11 @@ const typingBlackListFilter = createBlacklistFilter(
 
+const callsBlackListFilter = createBlacklistFilter(
+    'entities',
+    ['calls'],
```
done

jespino commented on app/products/calls/client/rest.ts:
```diff
@@ -0,0 +1,33 @@
+export interface ClientCallsMix {
+    getCalls: () => Promise<any>;
+    enableChannelCalls: (channelId: string) => Promise<any>;
+    disableChannelCalls: (channelId: string) => Promise<any>;
```
done

jespino commented on app/components/formatted_relative_time.tsx:
```diff
@@ -0,0 +1,41 @@
+    updateIntervalInSeconds: number;
+}
+
+const FormattedRelativeTime = ({timezone, value, updateIntervalInSeconds, ...props}: FormattedRelativeTimeProps) => {
```
Done here and in Call duration, now the updateIntervalInSeconds is optional.
